### PR TITLE
Select: add input filter

### PR DESCRIPTION
**Pull request by 3rdvision**
| - | - |
| --- | --- |
| Decisions | - Add `<input>` filter to `<select>` as now it makes sense to have this component considering all the use cases in which there are a lot of options. <br> - Based on https://github.com/ripe-tech/ripe-components-vue/pull/252 |
| Animated GIF | Peek 2021-05-27 12-46 |

### CODE REVIEW

File: vue/components/ui/molecules/select/select.stories.js [hunk 1]
```diff
@@ -14,39 +14,50 @@ storiesOf("Components/Molecules/Select", module)
...
+                    { value: "option_5", label: "Daniel" },
+                    { value: "option_6", label: "Gabriel" },
+                    { value: "option_61", label: "Gabriel11" },
+                    { value: "option_62", label: "Gabriel12" },
+                    { value: "option_63", label: "Gabriel3" },
+                    { value: "option_7", label: "Geraldo" },
+                    { value: "option_8", label: "João" },
+                    { value: "option_81", label: "João 1" },
+                    { value: "option_82", label: "João 2" },
+                    { value: "option_83", label: "João 3" },
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.stories.js:30]: Let's have some cool demo that: change the numbers to some random name.

File: vue/components/ui/molecules/select/select.vue [hunk 2]
```diff
@@ -264,11 +317,16 @@ export const Select = {
...
+            if (this.filter) {
+                this.filterText = "";
+                this.$nextTick(() => this.$refs.input.focus());
+            }
         },
         closeDropdown() {
             if (!this.visibleData) return;
             this.dehighlight();
             this.visibleData = false;
+            if (this.filter) this.$nextTick(() => this.$refs.selectButton.focus());
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:329]: Comment these

File: vue/components/ui/molecules/select/select.vue [hunk 3]
```diff
@@ -346,13 +410,14 @@ export const Select = {
             this.toggleDropdown();
         },
         onSelectButtonEnterKey() {
+            if (!this.filteredOptions[this.highlighted]) return;
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:413]: comment

File: vue/components/ui/molecules/select/select.vue [hunk 4]
```diff
@@ -346,13 +410,14 @@ export const Select = {
...
         onSelectButtonEnterKey() {
+            if (!this.filteredOptions[this.highlighted]) return;
             this.toggleDropdown();
         },
         onSelectButtonSpaceKey() {
             this.toggleDropdown();
         },
         onKey(key) {
-            if (!key) return;
+            if (!key || this.filter) return;
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:420]: comment

File: vue/components/ui/molecules/select/select.vue [hunk 5]
```diff
@@ -398,7 +463,11 @@ export const Select = {
                 return;
             }
 
-            this.setValue(this.options[this.highlighted].value);
+            if (!this.filteredOptions[this.highlighted]) {
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:466]: Why is this needed?

File: vue/components/ui/molecules/select/select.vue [hunk 6]
```diff
@@ -436,12 +509,22 @@ export const Select = {
...
+            if (!this.filter || !this.filterText) return this.options;
+            return this.options.filter(option => this.filterFunction(option, this.filterText));
+        },
         selectedOption() {
             return this.options ? this.options[this.valueIndex] : null;
         },
         buttonText() {
             return this.selectedOption ? this.selectedOption.label : this.placeholder;
         },
+        dropdownOwners() {
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:522]: Dead code?

**Review by gcandal — COMMENTED** [review]
Can you write unit tests?

File: vue/components/ui/molecules/select/select.vue (hunk 5 above)
Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue:466]: this is for the case in which the user presses enter but there's nothing highlighted in the list considering what's written in the select filter.

I personally found this to be the best way to handle the case - no action.

image

File: vue/components/ui/molecules/select/select.vue [hunk 7]
```diff
@@ -19,9 +19,27 @@
             </option>
         </select>
         <div class="select-container" v-bind:style="style" v-else>
+            <input-ripe
+                class="select-input"
+                v-bind:value.sync="filterText"
+                v-show="filter && visibleData"
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:25]: Move this to a computed, and use `!stuff` in the `v-show` of the select-button below, so it's clear that it's either one or the other.

**Review by gcandal — COMMENTED** [review]
Let's go over this in person as it has a ton of stuff in here.

File: vue/components/ui/molecules/select/select.vue [hunk 8]
```diff
@@ -122,10 +150,19 @@
...
 }
 
+.select .select-container .select-input:focus,
 .select .select-container .select-button:focus {
     border-color: $aqcua-blue;
 }
 
+.select.select-filter .select-container .select-button {
+    transition: width 0.2s ease, background-color 0.2s ease;
+}
```
[the remark this answers is not in the document]
Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue:160]: yes, to avoid a flicker in border color

File: vue/components/ui/molecules/select/select.vue [hunk 9]
```diff
@@ -380,11 +447,11 @@ export const Select = {
         },
         onAltUpKey() {
             this.openDropdown();
-            this.highlight(0);
+            this.highlight(0, true);
```
[the remark this answers is not in the document]
Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue:450]: yes, this 2 year bug still existed

File: vue/components/ui/molecules/select/select.vue [hunk 10]
```diff
@@ -263,11 +313,19 @@ export const Select = {
...
             this.visibleData = true;
+
+            // focus input and with blank text after dropdown is opened
+            if (this.filter) {
+                this.filterText = "";
+                this.$nextTick(() => this.$refs.input.focus());
+            }
         },
-        closeDropdown() {
+        closeDropdown(blur = false) {
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:323]: Do it in the positive (focus = true)

File: vue/components/ui/molecules/select/select.vue [hunk 11]
```diff
@@ -263,11 +313,19 @@ export const Select = {
...
+                this.$nextTick(() => this.$refs.input.focus());
+            }
         },
-        closeDropdown() {
+        closeDropdown(focus = true) {
             if (!this.visibleData) return;
             this.dehighlight();
             this.visibleData = false;
+            // focus the select button after dropdown is closed
+            if (this.filter && focus) this.$nextTick(() => this.$refs.selectButton.focus());
```
Comment (gcandal) [inline vue/components/ui/molecules/select/select.vue:328]: It seems weird that I can ask `closeDropdown(focus = true)` and then no focus happens because I don't have a `flter`. Having this said, you can achieve the same result and keep this function sane by moving the `this.filter` logic to the caller of this function

File: vue/components/ui/molecules/select/select.vue [hunk 12]
```diff
@@ -411,7 +483,12 @@ export const Select = {
                 return;
             }
 
-            this.setValue(this.options[this.highlighted].value);
+            // if there's no options for current filter, do nothing
+            if (!this.filteredOptions[this.highlighted]) {
```
[the remark this answers is not in the document]
Comment (gcandal) [reply vue/components/ui/molecules/select/select.vue:478]: Bump. I believe you can reuse from 477-482 in `onEnterKey`

File: vue/components/ui/molecules/select/select.vue (hunk 11 above)
Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue:328]: Focus is already happening by default if I don't have a `filter`. This is to fix the behavior of the filter because it would focus an element that would hide right after.

But there indeed a problem with this because when I do `closeDropdown(focus = false)` it will in fact focus in non-filter mode.

Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue:328]: refactoring..

File: vue/components/ui/molecules/select/select.vue (hunk 12 above)
[the remark this answers is not in the document]
Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue:478]: I remember we decided long ago to duplicated onEnterKey and onSpaceKey code [as it is in master ](https://github.com/ripe-tech/ripe-components-vue/blob/60e300cc47c5a7e1b6918c31ede756df0f7805e5/vue/components/ui/molecules/select/select.vue#L456)

simplifying this logic

File: vue/components/ui/molecules/select/select.vue (hunk 11 above)
Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue:328]: thinking better, there are a lot of places (8 places) that call closeDropdown and want `focus = false`
maybe it's better to just remove focus arg? as we will probably always want to focus the selectButton after closeDropdown

Comment (gcandal) [reply vue/components/ui/molecules/select/select.vue:328]: If you can always call focus without any issue then yes, I'd do that.

Comment (3rdvision) [reply vue/components/ui/molecules/select/select.vue]: I renamed to `maintainFocus` as removing the arg would stop global click from unfocusing.

Focus will always maintain itself in select for filter and non-filter mode.
The rename gives better info that if focus wasn't there in the beginning, it will not focus.

**Review by gcandal — APPROVED** [review]

**Comment by gcandal** [comment]
@3rdvision don't forget the changelog entry 🚀

**Comment by joamag** [comment]
Closing this until a proper rationale exists
